code/modules/hydroponics/hydro_tray_process.dm

```diff
@@ -0,0 +1,340 @@
+			health += healthmod
+		else
+			affect_growth(-1)
+			health -= healthmod
+	if(prob(35))
+		if(waterlevel < 10)
+			health += healthmod
+		else
+			affect_growth(-1)
+			health -= healthmod
+
+	// Check that pressure, heat and light are all within bounds.
+	// First, handle an open system or an unconnected closed system.
+
+	var/turf/T = loc
+	var/datum/gas_mixture/environment
+
+	// If we're closed, take from our internal sources.
+	if(closed_system && (connected_port || holding))
+		environment = air_contents
+
+	// If atmos input is not there, grab from turf.
+	if(!environment)
+		if(istype(T))
+			environment = T.return_air()
+
+	if(!environment)
+		if(istype(T, /turf/space))
+			environment = space_gas
+		else
+			return
+
+	// Handle gas consumption.
+	if(seed.consume_gasses && seed.consume_gasses.len)
+		missing_gas = 0
+		for(var/gas in seed.consume_gasses)
+			if(environment)
+				switch(gas)
+					if("oxygen")
+						if(environment.oxygen <= seed.consume_gasses[gas])
+							missing_gas++
+							continue
+					if("plasma")
+						if(environment.toxins >= seed.consume_gasses[gas])
+							missing_gas++
+							continue
+					if("nitrogen")
+						if(environment.nitrogen >= seed.consume_gasses[gas])
+							missing_gas++
+							continue
+					if("carbon_dioxide")
+						if(environment.carbon_dioxide >= seed.consume_gasses[gas])
+							missing_gas++
+							continue
+				environment.adjust_gas(gas,-seed.consume_gasses[gas],1)
+			else
+				missing_gas++
+
+		if(missing_gas > 0)
+			health -= missing_gas * HYDRO_SPEED_MULTIPLIER
+
+	// Process it.
+	var/pressure = environment.return_pressure()
+	if(pressure < seed.lowkpa_tolerance || pressure > seed.highkpa_tolerance)
+		health -= healthmod
+		improper_kpa = 1
+	else
+		improper_kpa = 0
+
+	if(abs(environment.temperature - seed.ideal_heat) > seed.heat_tolerance)
+		health -= healthmod
+		improper_heat = 1
+	else
+		improper_heat = 0
+
+	// Handle gas production.
+	if(seed.exude_gasses && seed.exude_gasses.len)
+		for(var/gas in seed.exude_gasses)
+			environment.adjust_gas(gas, max(1,round((seed.exude_gasses[gas]*round(seed.potency))/seed.exude_gasses.len)))
+
+	// This was adapted from the air alarm code
+	// I've never done atmos or touched atmos code before so there's a chance I've fucked this up
+	if(seed.alter_temp)
+		if(istype(T, /turf/simulated/))
+			var/datum/gas_mixture/room = T.remove_air(environment.total_moles)
+			if(room.temperature < seed.ideal_heat-seed.heat_tolerance)
+				room.temperature += (seed.potency*3)
+			else if (room.temperature > seed.ideal_heat+seed.heat_tolerance)
+				room.temperature -= (seed.potency*3)
+			room.react()
+			T.assume_air(room)
+
+	// If we're attached to a pipenet, then we should let the pipenet know we might have modified some gasses
+	//if (closed_system && connected_port)
+	//'	update_connected_network()
+
+	// Handle light requirements.
+
+	var/light_available = 5
+	if(T.dynamic_lighting)
+		light_available = T.get_lumcount() * 10
```

Comment:
So, 1 lumens of light is actually 10 of them?

`/turf/proc/get_lumcount(var/minlum, var/maxlum)` returns a scaled lumcount between `minlum` and `maxlum`; this means that doing `get_lumcount(5, 10)` will make it so that anything less than `5` is treated as `0`, and anything bigger than `10` is capped to `10`, 7.5 lums would return `0.5`...